### PR TITLE
Implemented working pretrain stage

### PR DESCRIPTION
**Pull request by NonameUntitled**
Command to run: ```python pretrain.py --params configs\pretrain_config.json``` from project directory

What was implemented in detail:

- Example of how to connect topology and ML-related algorithms
- Tensorboard utils for quality monitoring. Command to run: ```tensorboard --logdir=tensorboard_logs --reload_multifile=TRUE``` from project dir (don't forget to create `tensorboard_logs ` dir)
- Topology parsing from config
- Learnable embeddings
- Shared embeddings support
- Utils
- A lot of TODO's with ideas and next steps

What wasn't done:
- Typing for all the functions
- Documentation (probably it is better to start investing time in it)
- Detailed debugging of the pretraining stage (It is also really important)
- Correct imports and function argument order (Have to be discussed and standardized)

### CODE REVIEW

File: agents/dqn.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 from ml.encoders import BaseEncoder
 
 
+# TODO[Vladimir Baikalov]: Implement
```
Comment (zhog96) [inline agents/dqn.py:5]: I started DQN in another branch, will refactor like PPO in this branch =)

File: topology/base.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+                snd_node_cfg = conveyor_section[i]
+
+                fst_node_position = fst_node_cfg.position
+                snd_node_position = snd_node_cfg.position
+
+                edge_length = snd_node_position - fst_node_position
+
+                assert edge_length >= 2, \
+                    f'Conveyor section of conveyor {conveyor_id} is too short. ' \
+                    f'Positions: {fst_node_position} and {fst_node_position}!'
```
[the remark this answers is not in the document]
Comment (NonameUntitled) [reply topology/base.py:130]: As far as I remember, it was something about bag sizes and connection to real-world conveyors. In the previous implementation bags have sizes so I assume it was done just to make it more realistic (Idk frankly speaking)

File: topology/base.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+                    conveyor=conveyor_id,
+                    end_pos=snd_node_position
+                )
+
+                if (
+                        i > 1  # This node stays on this conveyor (at some point)
+                        or fst_node_cfg.type != 'diverter'  # This conveyor goes from source
+                ):
+                    graph.nodes[fst_node_cfg]['conveyor'] = conveyor_id
+                    graph.nodes[fst_node_cfg]['conveyor_pos'] = fst_node_position
```
[the remark this answers is not in the document]
Comment (NonameUntitled) [reply topology/base.py:145]: For the first node it is incorrect. If the first node is a diverter, then then it is located on the other conveyor so we simply can't broadcast this info along all the nodes. There is only one exception (second condition) if the first node in the list is the source, then it simply doesn't have any conveyor related to its location, so we have only one option to assign this node to the new conveyor.

File: ml/losses.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    pass
+
+
+class IdentityLoss(BaseLoss, config_name='identity'):
+
+    def __call__(self, inputs):
+        return inputs
+
+
+class CompositeLoss(TorchLoss, config_name='composite'):
```
[the remark this answers is not in the document]
Comment (NonameUntitled) [reply ml/losses.py:21]: It is just a utility class to combine losses from above. Maybe in future we will somehow use it for train stage as well, but for now it is done just for pretrain, you  don't need to embed it into your pipeline

File: agents/ppo.py [hunk 5]
```diff
@@ -4,110 +4,116 @@
...
+            destination_node_idx=destination_node_idx
         )
-        return next_node_idx
+
+        if self._bag_trajectory_memory is not None:
+            # TODO[Vladimir Baikalov]: Think about how to generalize
+            # Shape: [batch_size] if exists, None otherwise
+            bag_ids = inputs.get(self._bag_idx_prefix, None)
+            # Static entity if exists, None otherwise
+            storage = inputs.get(self._storage_prefix, None)
```
[the remark this answers is not in the document]
Comment (NonameUntitled) [reply agents/ppo.py:96]: It is still needed because we use it below to save trajectory in the memory. It is correct for `bag_ids`, I will remove `storage` since it's unused, thanks

[the remark this answers is not in the document]
Comment (NonameUntitled) [reply agents/ppo.py:96]: Yes, you are right. I will remove `storage` but still we need `bag_ids` for storing the trajectory

File: ml/ppo_encoders.py [hunk 6]
```diff
@@ -41,31 +44,82 @@ def __init__(self, embedder: TorchEncoder, ff_net: TorchEncoder):
...
+        # Shape: [batch_size, max_neighbors_num]
+        neighbors_probs = torch.nn.functional.softmax(neighbors_logits, dim=1)
+        neighbors_probs[~neighbor_node_embeddings.mask] = 0
 
-        neighbour_logits = euclidean_distance(neighbour_embs, destination_embs, dim=1)
-        neighbour_probs = softmax(TensorWithMask(1 / (neighbour_logits.tensor + 1e-18), neighbour_logits.mask), dim=1)
-        next_neighbour = sample(neighbour.tensor, neighbour_probs)
-        return next_neighbour, neighbour_probs
+        # TODO[Vladimir Baikalov]: Check these two lines below
+        neighbors_probs /= neighbors_probs.max()
```
[the remark this answers is not in the document]
Comment (NonameUntitled) [reply ml/ppo_encoders.py:103]: Oh, my bad. I don't know what I was thinking about. Have to fix it, thanks.

File: ml/ppo_encoders.py [hunk 7]
```diff
@@ -41,31 +44,82 @@ def __init__(self, embedder: TorchEncoder, ff_net: TorchEncoder):
...
+        # 3) Compute logits for existing neighbors
+        # Shape: [batch_size, max_neighbors_num]
+        neighbors_logits = torch.einsum(
+            'bnd,bd->bn',
+            shifted_padded_neighbors_node_embeddings,
+            ideal_transition_embedding
+        )
+        # TODO[Vladimir Baikalov]: Probably it's a good idea to divide logits to make the distribution more smooth
+        # TODO[Vladimir Baikalov]: Put constant in the variable
+        neighbors_logits[~neighbor_node_embeddings.mask] = -1e18
```
[the remark this answers is not in the document]
Comment (NonameUntitled) [reply ml/ppo_encoders.py:95]: That's a good point, I will examine how to make it work in a proper way. Initially, I just wrote in the way to make it work

File: ml/losses.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        for loss, weight in zip(self._losses, self._weights):
+            total_loss += weight * loss(inputs)
+
+        if self._output_prefix is not None:
+            inputs[self._output_prefix] = total_loss.cpu().item()
+
+        return total_loss
+
+
+class CrossEntropyLoss(TorchLoss, config_name='cross_entropy'):
```
[the remark this answers is not in the document]
Comment (NonameUntitled) [reply ml/losses.py:54]: For the train loss there is a custom function inside the PPOAgent, you don't need to follow this example, it was done just to simplify pretrain stage configuration. For the train stage, you have to write loss in the PPOActor which will be invoked in the system so it is no need in using utilities from this directory.

**Review by zhog96 — APPROVED** [review]
